### ISSUE TITLE
upstream connect error or disconnect/reset before headers. reset reason: connection failure

### ISSUE THREAD

**Issue by cristianmtr** [cortexlabs/cortex#1269]
#### Version

cli version: 0.18.1

#### Description

Intermittent 503 errors on AWS cluster.

#### Configuration

cortex.yaml

```
# cortex.yaml

- name: offer-features
  predictor:
    type: python
    path: predictor.py
    config:
      bucket: XXXXXXXXXXXXXXXXXXXX
  compute:
    cpu: 1  # CPU request per replica, e.g. 200m or 1 (200m is equivalent to 0.2) (default: 200m)
    gpu: 0  # GPU request per replica (default: 0)
    inf: 0 # Inferentia ASIC request per replica (default: 0)
    mem: 1Gi
  autoscaling:
    min_replicas: 2
    max_replicas: 3
    init_replicas: 2
    max_replica_concurrency: 13
    target_replica_concurrency: 5
    window: 1m0s
    downscale_stabilization_period: 5m0s
    upscale_stabilization_period: 1m0s
    max_downscale_factor: 0.75
    max_upscale_factor: 1.5
    downscale_tolerance: 0.05
    upscale_tolerance: 0.05
```

```
# cluster.yaml

# AWS credentials (if not specified, ~/.aws/credentials will be checked) (can be overridden by $AWS_ACCESS_KEY_ID and $AWS_SECRET_ACCESS_KEY)
aws_access_key_id: XXXXXXXXXXXXXX
aws_secret_access_key: XXXXXXXXXXXXXXXXX

# optional AWS credentials for the operator which may be used to restrict its AWS access (defaults to the AWS credentials set above)
cortex_aws_access_key_id: XXXXXXXXXXXXXXXX
cortex_aws_secret_access_key: XXXXXXXXXXXXXXXXXXXXX

# EKS cluster name for cortex (default: cortex)
cluster_name: cortex

# AWS region
region: us-east-1

# S3 bucket (default: <cluster_name>-<RANDOM_ID>)
# note: your cortex cluster uses this bucket for metadata storage, and it should not be accessed directly (a separate bucket should be used for your models)
bucket: # cortex-<RANDOM_ID>

# list of availability zones for your region (default: 3 random availability zones from the specified region)
availability_zones: # e.g. [us-east-1a, us-east-1b, us-east-1c]

# instance type
instance_type: t3.medium

# minimum number of instances (must be >= 0)
min_instances: 1

# maximum number of instances (must be >= 1)
max_instances: 5

# disk storage size per instance (GB) (default: 50)
instance_volume_size: 50

# instance volume type [gp2, io1, st1, sc1] (default: gp2)
instance_volume_type: gp2

# instance volume iops (only applicable to io1 storage type) (default: 3000)
# instance_volume_iops: 3000

# whether the subnets used for EC2 instances should be public or private (default: "public")
# if "public", instances will be assigned public IP addresses; if "private", instances won't have public IPs and a NAT gateway will be created to allow outgoing network requests
# see https://docs.cortex.dev/v/0.18/miscellaneous/security#private-cluster for more information
subnet_visibility: public  # must be "public" or "private"

# whether to include a NAT gateway with the cluster (a NAT gateway is necessary when using private subnets)
# default value is "none" if subnet_visibility is set to "public"; "single" if subnet_visibility is "private"
nat_gateway: none  # must be "none", "single", or "highly_available" (highly_available means one NAT gateway per availability zone)

# whether the API load balancer should be internet-facing or internal (default: "internet-facing")
# note: if using "internal", APIs will still be accessible via the public API Gateway endpoint unless you also disable API Gateway in your API's configuration (if you do that, you must configure VPC Peering to connect to your APIs)
# see https://docs.cortex.dev/v/0.18/miscellaneous/security#private-cluster for more information
api_load_balancer_scheme: internet-facing  # must be "internet-facing" or "internal"

# whether the operator load balancer should be internet-facing or internal (default: "internet-facing")
# note: if using "internal", you must configure VPC Peering to connect your CLI to your cluster operator (https://docs.cortex.dev/v/0.18/guides/vpc-peering)
# see https://docs.cortex.dev/v/0.18/miscellaneous/security#private-cluster for more information
operator_load_balancer_scheme: internet-facing  # must be "internet-facing" or "internal"

# CloudWatch log group for cortex (default: <cluster_name>)
log_group: cortex

# additional tags to assign to aws resources for labelling and cost allocation (by default, all resources will be tagged with cortex.dev/cluster-name=<cluster_name>)
tags:  # <string>: <string> map of key/value pairs

# whether to use spot instances in the cluster (default: false)
# see https://docs.cortex.dev/v/0.18/cluster-management/spot-instances for additional details on spot configuration
spot: false

# see https://docs.cortex.dev/v/0.18/guides/custom-domain for instructions on how to set up a custom domain
ssl_certificate_arn: XXXXXXXXXXXXXXXXXXXXXXXXXXXX

```

#### Steps to reproduce

- Spin up instances on AWS.
- Wait a couple of days / hours (varies).
- Notice sudden 503 errors

#### Expected behavior

It should work

#### Actual behavior

503 errors with the message

```
upstream connect error or disconnect/reset before headers. reset reason: connection failure
```

#### Screenshots

NOTE: The endpoint stopped responding around 15:30 in the graphs below.

Monitoring nr of bytes in:

![image](https://user-images.githubusercontent.com/8330330/90371369-267ded00-e06f-11ea-8980-05aaac13f98f.png)

Nr of requests:

![image](https://user-images.githubusercontent.com/8330330/90377607-7e6d2180-e078-11ea-8018-1267f5ddb8a8.png)

#### Stack traces

Nothing useful, just:

```text
2020-08-16 05:38:34.697979:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:38:37.643022:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:38:40.577522:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:38:42.008412:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:38:43.513294:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:38:45.425255:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:38:48.327276:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:38:51.316962:cortex:pid-447:INFO:200 OK POST /predict
2020-08-16 05:38:54.009212:cortex:pid-447:INFO:200 OK POST /predict
2020-08-16 05:38:55.852878:cortex:pid-447:INFO:200 OK POST /predict
2020-08-16 05:38:57.525264:cortex:pid-447:INFO:200 OK POST /predict
2020-08-16 05:39:00.795236:cortex:pid-447:INFO:200 OK POST /predict
2020-08-16 05:39:04.437013:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:39:05.981920:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:39:09.314293:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:39:12.343143:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:39:15.821708:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:39:19.083554:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:39:22.048843:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:39:24.943968:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:39:26.613330:cortex:pid-448:INFO:200 OK POST /predict
2020-08-16 05:39:29.702703:cortex:pid-448:INFO:200 OK POST /predict

```

#### Additional context

- The prediction takes about ~150ms on my Dell with Intel© Core™ i7-8750H CPU @ 2.20GHz × 6, 32GB Ram.
- All the load balancer target are marked as "unhealthy", even when they work (i.e. I can send requests and receive 2XX responses)
- The load balancer healthcheck endpoint returns the following

```
/healthz
{
        "service": {
                "namespace": "istio-system",
                "name": "ingressgateway-operator"
        },
        "localEndpoints": 0
}%  
```

#### Suggested solution

(optional)

**Comment by RobertLucian** [comment]
@cristianmtr thanks for reporting this here.

We already have a ticket for it #814.
This may be valuable in debugging this issue so we're gonna keep this one open.

**Comment by cristianmtr** [comment]
@RobertLucian Thanks for answering. Are you also seeing the "unhealthy" status on the target instances for the load balancer? Or is that a separate issue I should look into?

**Comment by RobertLucian** [comment]
@cristianmtr we haven't tested for that yet, but we will look into this as we investigate #814 to see if it's related. Thanks for the heads up. We'll dig into this soon and update you on this matter.

Also, we do have a couple of questions regarding this:
1. How often does this surge in 503s happen to you? Every hour? Every few days?
1. When this happens, what percentage of the total number of requests get 503s?
1. When this happens, how long would you say this persists for? 1 minute, 5 minutes, 1 hour, etc?
1. Are you using the API Gateway? The API Gateway is enabled by default and its field is `networking:api_gateway` in `cortex.yaml`.
1. When you say redeployment, do you mean `cortex delete <api-name> && cortex deploy` or just `cortex deploy`?

**Comment by cristianmtr** [comment]
> How often does this surge in 503s happen to you? Every hour? Every few days?
Last time it happened it was after 2 days since the deployment.
> When this happens, what percentage of the total number of requests get 503s?
It's a complete shutdown. No requests reach the instances themselves. Our theory is that they are stopped at the LB level (due to the 'unhealthy' issue I mentioned)
> When this happens, how long would you say this persists for? 1 minute, 5 minutes, 1 hour, etc?
I've seen it twice when it automatically "fixed" itself almost immediately. Otherwise it's permanent.
> Are you using the API Gateway? The API Gateway is enabled by default and its field is networking:api_gateway in cortex.yaml.
Yes, we are. We have also enabled SSL at that level, if that helps.
> When you say redeployment, do you mean cortex delete <api-name> && cortex deploy or just cortex deploy?
It's enough to `deploy` (though we need to make a dummy change to be able to, otherwise we are told the API is up to date)

**Comment by cristianmtr** [comment]
I investigated the error message from the load balancer health checks, and I ended up at this issue on the kubernetes github page: https://github.com/kubernetes/kubernetes/issues/61486

Seems like it's a problem with the VPC setup in AWS. Not sure this is related to the problem you referenced above.

**Comment by RobertLucian** [comment]
@cristianmtr we have reasons to think that this is no longer related to #814 (just like you have mentioned in one of your previous responses).

Our course of action would be to first reproduce this. In this case, could you tell us all of the steps you've taken so far that got you to this point (or at least to the first time when you encountered this)? (i.e. installed CLI, spun-up the cluster, etc). The more details there are, the faster/easier should be for us to reproduce this. Anything you think is useful, please let us know.

**Comment by cristianmtr** [comment]
@RobertLucian Thanks for looking into this

- installed CLI
- implemented Python predictor

A few notes here:

- in the `__init__` the class downloads and loads three fairly large models (based on Spacy)
- a request takes about 350ms to process on a `t3.medium`

- spun up cluster using the config above
- configured the API Gateway to serve using SSL. Added route to Route53
- the endpoint gets ~30 req/min from our services. The data is just text, and the payload size isn't large
- the service returns `503 upstream connect error or disconnect/reset before headers. reset reason: connection failure` after a random amount of time (we haven't been able to pinpoint what's the underlying reason)
- a `refresh` or `deploy` (with a dummy change) always fixes the issue

Lmk if you need more specific details

**Comment by deliahu** [comment]
@cristianmtr in our [v0.22 release](https://github.com/cortexlabs/cortex/releases/tag/v0.22.0) today, we redesigned part of our networking stack. We were able to confirm that it fixed https://github.com/cortexlabs/cortex/issues/814. I know that this ticket is unrelated, but it's possible that this issue is also resolved by v0.22 (we have not been able to reproduce it, although we weren't able to reproduce it on v0.21 either).

Have you continued to experience the 503s, and if so, do you mind trying on cortex v0.22?

**Comment by deliahu** [comment]
@cristianmtr I just wanted to follow up on this; has this error been persisting, and if so, have you tried on v0.22+?

**Comment by cristianmtr** [comment]
@deliahu Sorry, I no longer work at that company. And even before I left, we had switched over from cortex unfortunately. to AWS EB

**Comment by RobertLucian** [comment]
@cristianmtr understandable. Since you were in the company at the time when you decided to switch away from Cortex, it would help us a lot if you could share some feedback about Cortex:

* Why did you transition from Cortex to AWS EB?
* What features did AWS EB offer that Cortex didn't?
* What would have kept you with Cortex?
* What things do you think we should address?
* What do you personally feel/think of Cortex?

I feel like this is an opportunity for us to improve by hearing your reasons. You'd be helping us a lot!

**Comment by cristianmtr** [comment]
Mainly because of this intermittent problem, and we had already used EB for serving some of our models/services.

@nbeuchat

**Comment by deliahu** [comment]
I believe this has been resolved as of version 0.22. Please let us know if this problem persists.
